packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -461,6 +500,82 @@ class RouteSettings {
   String toString() => '${objectRuntimeType(this, 'RouteSettings')}("$name", $arguments)';
 }
 
+/// Describes the configuration of a [Route].
+///
+/// The type argument `T` is the corresponding [Route]'s return type, as
+/// used by [Route.currentResult], [Route.popped], and [Route.didPop].
+///
+/// See also:
+///
+///  * [Navigator.pages], which accepts a list of [Page]s and updates its routes
+///    history.
+///  * [PageBuilder], a [Page] subclass that provides the API to build a
+///    customized route.
+abstract class Page<T> extends RouteSettings {
+  /// Creates a page and initializes [key] for subclasses.
+  ///
+  /// The [arguments] argument must not be null.
+  const Page({
+    this.key,
+    String name,
+    Object arguments,
+  }) : super(name: name, arguments: arguments);
+
+  /// The key associates with this page.
+  ///
+  /// This key will be used for comparing pages in [canUpdate].
+  final LocalKey key;
+
+  /// Whether this page can be updated with the [other] page.
+  ///
+  /// Two pages are consider updatable if they have same the [runtimeType] and
+  /// [key].
+  bool canUpdate(Page<dynamic> other) {
+    return other.runtimeType == runtimeType &&
+      other.key == key;
+  }
+
+  /// Creates the [Route] that corresponds to this page.
+  ///
+  /// The created [Route] must have its [Route.settings] property set to this [Page].
+  Route<T> createRoute(BuildContext context);
+
+  @override
+  String toString() => '${objectRuntimeType(this, 'Page')}("$name", $key, $arguments)';
+}
+
+/// A [Page] that builds a customized [Route] based on the [routeBuilder].
+///
+/// The type argument `T` is the corresponding [Route]'s return type, as
+/// used by [Route.currentResult], [Route.popped], and [Route.didPop].
+class PageBuilder<T> extends Page<T> {
```

Comment:
I am thinking how useful is this builder. We are not surfacing page api, and it is likely the Router will take care of changing route name string to configured navigator.